File: frontend/src/lib/session.ts
// Session resilience utilities for the merchant dashboard.
//
// The dashboard is a long-lived SPA that authenticates via Frappe's `sid`
// cookie and a CSRF token (`window.csrf_token`) captured once at page load.
// When the server-side session ends or the CSRF token goes stale, Frappe
// clears the cookie and the SPA — which trusts a frozen boot snapshot — would
// otherwise dump the user at /login with no warning.
//
// This module is the shared toolkit the SessionGuard uses to:
//   • refresh a stale CSRF token without a full reload (self-heal writes)
//   • probe liveness for the heartbeat
//   • broadcast a single "session expired" signal to the UI
//   • beacon diagnostics so production reveals the real trigger

const METHOD = 'flamezo_backend.flamezo.api.session'

/** Full (non-SPA) login URL — used for hard redirects out of the app. */
export const LOGIN_URL = '/flamezo_backend/login'

const CSRF_PLACEHOLDER = '{{ csrf_token }}'

/** The user baked into the page at load time (object or string, or Guest). */
export function getBootUser(): string {
  const raw = (window as any)?.frappe?.boot?.user
  const name = typeof raw === 'object' && raw != null ? raw.name : raw
  return name || 'Guest'
}

/** True when the page loaded with a real (non-Guest) authenticated user. */
export function bootedAuthenticated(): boolean {
  const u = getBootUser()
  return !!u && u !== 'Guest'
}

/** Current CSRF token the SDK will send, or '' if unusable. */
export function currentCsrfToken(): string {
  const t = (window as any)?.csrf_token
  return t && t !== CSRF_PLACEHOLDER ? String(t) : ''
}

function setCsrfToken(token: string) {
  if (token) (window as any).csrf_token = token
}

interface CsrfResult {
  authenticated: boolean
  csrf_token: string
}

/**
 * Ask the server for a fresh CSRF token and update `window.csrf_token`.
 * The SDK reads `window.csrf_token` on every request, so updating it here is
 * enough for subsequent (and retried) writes to carry the correct token.
 * Returns `authenticated: false` when the session is already gone.
 */
export async function refreshCsrfToken(): Promise<CsrfResult> {
  try {
    const res = await fetch(`/api/method/${METHOD}.get_csrf_token`, {
      method: 'GET',
      credentials: 'include',
      cache: 'no-store',
      headers: { Accept: 'application/json' },
    })
    const json = await res.json().catch(() => ({}))
    const data = json?.message ?? json
    const token = data?.csrf_token || ''
    if (token) setCsrfToken(token)
    return { authenticated: !!data?.authenticated, csrf_token: token }
  } catch {
    return { authenticated: false, csrf_token: '' }
  }
}

/** Cheap heartbeat probe. Never throws; reports live auth state. */
export async function pingSession(): Promise<CsrfResult> {
  try {
    const res = await fetch(`/api/method/${METHOD}.ping`, {
      method: 'GET',
      credentials: 'include',
      cache: 'no-store',
      headers: { Accept: 'application/json' },
    })
    const json = await res.json().catch(() => ({}))
    const data = json?.message ?? json
    const token = data?.csrf_token || ''
    if (token) setCsrfToken(token)
    return { authenticated: !!data?.authenticated, csrf_token: token }
  } catch {
    // Network blip — assume still alive; the next real request decides.
    return { authenticated: true, csrf_token: currentCsrfToken() }
  }
}

export interface SessionLossContext {
  reason: string
  url?: string
  method?: string
  status?: number
  responseSnippet?: string
  hadSidCookie?: boolean
  bootUser?: string
  at?: string
}

/** Was a `sid` cookie present at this moment? (HttpOnly hides the real one, */
/** but a duplicate non-HttpOnly `sid` — the classic logout cause — shows up.) */
export function readableSidCookiePresent(): boolean {
  return document.cookie.split(';').some((c) => c.trim().startsWith('sid='))
}

/**
 * Beacon the captured state at a forced logout so production logs the real
 * trigger. Uses sendBeacon (survives page unload) and falls back to fetch.
 */
export function reportSessionLoss(ctx: SessionLossContext): void {
  try {
    const payload = JSON.stringify({
      context: {
        ...ctx,
        bootUser: ctx.bootUser ?? getBootUser(),
        hadSidCookie: ctx.hadSidCookie ?? readableSidCookiePresent(),
        at: ctx.at ?? new Date().toISOString(),
        userAgent: navigator.userAgent,
        path: window.location.pathname,
      },
    })
    const url = `/api/method/${METHOD}.log_session_diagnostic`
    const body = new Blob([payload], { type: 'application/json' })
    if (navigator.sendBeacon && navigator.sendBeacon(url, body)) return
    fetch(url, { method: 'POST', credentials: 'include', body: payload, headers: { 'Content-Type': 'application/json', Accept: 'application/json' }, keepalive: true }).catch(() => {})
  } catch {
    // Diagnostics are best-effort.
  }
}

// ─── Single "session expired" signal ────────────────────────────────────────
// One source of truth so the whole app reacts once (modal), not N times.

let sessionExpired = false
const listeners = new Set<() => void>()

export function onSessionExpired(cb: () => void): () => void {
  listeners.add(cb)
  return () => listeners.delete(cb)
}

export function isSessionExpired(): boolean {
  return sessionExpired
}

/** Fire the session-expired signal exactly once. */
export function triggerSessionExpired(ctx: SessionLossContext): void {
  if (sessionExpired) return
  sessionExpired = true
  reportSessionLoss(ctx)
  listeners.forEach((cb) => {
    try {
      cb()
    } catch {
      /* listener errors must not cascade */
    }
  })
}

/** Reset (used after a successful re-auth in dev/tests). */
export function clearSessionExpired(): void {
  sessionExpired = false
}

// ─── Error classification ───────────────────────────────────────────────────

/** A stale/missing CSRF token — recoverable by refreshing + retrying once. */
export function isCsrfError(err: any): boolean {
  const status = err?.response?.status
  if (status !== 400 && status !== 403) return false
  const data = err?.response?.data
  const excType = String(data?.exc_type || data?.exception || '')
  // Frappe reliably tags CSRF rejections with this exc_type — prefer it so we
  // never mistake a PermissionError (also 403) for a recoverable CSRF failure.
  if (/CSRFToken/i.test(excType)) return true
  // Fallback only when no exc_type is present (e.g. a bare proxy/CSRF reject).
  if (!excType) {
    const messages = JSON.stringify(data?._server_messages || data?.message || '')
    return /csrf/i.test(messages)
  }
  return false
}

/** A genuinely gone session — not recoverable without re-login. */
export function isSessionGoneError(err: any): boolean {
  const status = err?.response?.status
  const data = err?.response?.data
  if (status === 401) return true
  // Frappe sets this on the response when it clears the session.
  if (data?.session_expired === 1 || data?.session_expired === '1') return true
  const excType = data?.exc_type || data?.exception || ''
  return /(AuthenticationError|SessionExpired)/i.test(excType)
}

File: flamezo_backend/flamezo/api/session.py
# Copyright (c) 2025, Flamezo and contributors
# For license information, please see license.txt

"""
Session lifecycle endpoints for the merchant dashboard SPA.

These power the client-side SessionGuard which keeps long-lived dashboard
sessions resilient:

  • get_csrf_token        — hand the SPA a fresh CSRF token so a stale
                            `window.csrf_token` (frozen at page load) can be
                            refreshed without a full reload. Self-heals the
                            403 / "Invalid Request" failures that otherwise
                            surface on write actions.

  • ping                  — cheap "am I still logged in?" probe for the
                            heartbeat. Returns the live user without throwing
                            so the client can decide how to react.

  • log_session_diagnostic — capture the exact state at the moment the SPA
                            detects a forced logout (cookies present, sid,
                            last request, response) so production tells us the
                            real trigger instead of us guessing.
"""

import json

import frappe
import frappe.sessions
from frappe import _


@frappe.whitelist(allow_guest=True)
def get_csrf_token():
	"""Return the current session's CSRF token.

	Safe to call as a GET (no CSRF required for GET), which is exactly why the
	SPA can use it to recover from a stale token. For a Guest the token is
	empty — the client treats that as "session gone" and surfaces re-login.
	"""
	token = ""
	user = frappe.session.user if frappe.session else "Guest"

	if user and user != "Guest":
		# Generates-and-persists on first call, returns the existing one after.
		token = frappe.sessions.get_csrf_token()
		frappe.db.commit()

	return {
		"csrf_token": token,
		"user": user,
		"authenticated": bool(user and user != "Guest"),
	}


@frappe.whitelist(allow_guest=True)
def ping():
	"""Lightweight liveness probe for the heartbeat.

	Never throws on an expired session — returns ``authenticated: False`` so the
	client can drive the graceful re-login flow instead of eating an exception.
	"""
	user = frappe.session.user if frappe.session else "Guest"
	authenticated = bool(user and user != "Guest")

	# Read-only: never generate a token here (generation writes to the session
	# and needs a live session object). Echo the existing one if present so the
	# client can re-assert it after a silent rotation; otherwise empty.
	existing_token = ""
	if authenticated:
		try:
			existing_token = frappe.session.data.csrf_token or ""
		except (AttributeError, KeyError):
			existing_token = ""

	return {
		"authenticated": authenticated,
		"user": user if authenticated else None,
		"csrf_token": existing_token,
	}


@frappe.whitelist(allow_guest=True)
def log_session_diagnostic(context=None):
	"""Record the client-side state captured at a forced-logout event.

	The SPA posts a small JSON blob (last request, response status, whether a
	`sid` cookie was present, timestamp, url). We persist it to the Error Log so
	we can correlate real merchant logouts with server behaviour and pin the
	exact trigger. Best-effort: never raises back to the client.
	"""
	try:
		if isinstance(context, str):
			try:
				context = json.loads(context)
			except (ValueError, TypeError):
				context = {"raw": context}

		payload = {
			"reported_user": (frappe.session.user if frappe.session else None),
			"request_ip": frappe.local.request_ip if hasattr(frappe.local, "request_ip") else None,
			"user_agent": frappe.get_request_header("User-Agent"),
			"client": context or {},
		}

		frappe.log_error(
			title="SPA Session Diagnostic",
			message=frappe.as_json(payload, indent=2),
		)
	except Exception:
		# Diagnostics must never break the client's logout flow.
		pass

	return {"logged": True}
